### PR TITLE
Add .x resolution and --resolve command

### LINKED ISSUE #110: gimme awareness of available versions
At present, the `gimme` script knows very little about what versions of Go exist and are potentially available.  There's testdata, which has that.  At a layer _above_ `gimme`, travis-ci/travis-build is doing some substitution work to make `1.9.x` work, by knowing what `.x` should be at any given point in time.

People keep expecting this to be part of travis-ci/gimme itself, which is absolutely fair.  It looks like `gimme`, it smells like `gimme`, they're filing bugs against `gimme` and it would be _useful_ for `gimme` to handle this.

Should travis-ci/gimme handle this?

1. No, keep things as they are
2. Detect `.x` and error out with a diagnostic explicitly pointing towards something higher up, suggesting travis-build.
3. Implement version tracking and caching as part of `gimme` itself, or perhaps just ability to spot in-repo data when run from a repo?
4. Something else?

### LINKED ISSUE #129: stable cache will not refresh properly on non-Linux
[While writing a new feature, I spotted a flaw; filing a bug for tracking independently of the new feature.  I expect to close this soon if not interrupted.]

The `_stat_unix` function uses modification time on Unix, but last _access_ time on macOS and BSDs.

Thus repeated accesses of `stable` will never result in `stable` being checked to be current.

### PR DESCRIPTION
**Pull request by philpennock**
* New persistent file, `~/.gimme/versions/known-versions.txt`, which is
  kept in sorted unique order
* Add `--resolve` for handling version specifiers in a
  normally-lightweight manner, where stdout gets the resolved version.
* Handle `NN.x` (and `NN.MM.x`) as a version specifier, against the
  known versions
* Have the `--known` output be cached to disk; there's a cache bypass
  mechanism (flag), and a cache age control environment variable.
* Rework some cache handling stuff to be in cleaner functional
  abstractions
* Adds a `_version_sort` filter
* Bug-fix for stable age using access-time not mod-time on darwin|*bsd

See the README updates for notes on interaction between `--resolve` and
git tag resolution, and how to avoid that.

Fixes #129
Resolves #110

### CODE REVIEW

**Comment by philpennock** [comment]
nb: I actually want to resolve this into master, but since it adds to the documentation added by #128 I've based the review diff off of that.

**Comment by philpennock** [comment]
~Re travis build failures: I disbelieve that `./gimme -h` can stall for 10 minutes, as a result of anything in this PR, or at all.~  I was wrong, I messed up.

**Comment by philpennock** [comment]
NB: Do not merge until #131 is approved and merged into this; GitHub handles stacked PRs very badly.

File: gimme
Comment (AlekSi) [inline gimme:22]: I think it is worth mentioning here that version _must_ be given as an argument. `GIMME_GO_VERSION` will not work.

File: README.md
Comment (AlekSi) [inline README.md]: Is it `--force-known-update`?

**Review by AlekSi — COMMENTED** [review]
I don't know bash that well, so I mostly checked the documentation.

Thank you for working on this!

**Comment by AlekSi** [comment]
Does not work in Raspbian:
```bash
pi@raspberrypi:~/gimme $ bash --version
GNU bash, version 4.4.12(1)-release (arm-unknown-linux-gnueabihf)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

pi@raspberrypi:~/gimme $ ./gimme -k
./gimme: line 478: /home/pi/.gimme/versions/known-versions.txt.new: No such file or directory
mv: cannot stat '/home/pi/.gimme/versions/known-versions.txt.new': No such file or directory
cat: /home/pi/.gimme/versions/known-versions.txt: No such file or directory
```

Here is a trace: https://gist.github.com/AlekSi/13afc22c0d62008118f5633c8341f9a4

File: gimme
Comment (AlekSi) [inline gimme:107]: Raspbian 9.3 has it. Probably, that's why it does not work on it.

File: README.md
Comment (philpennock) [reply README.md]: fixed, thanks

File: gimme
Comment (philpennock) [reply gimme:22]: oops, good call; I changed the code to handle `GIMME_GO_VERSION` instead though

**Comment by philpennock** [comment]
Identified the issues with Raspbian above: it's not Raspbian.  It's that `~/.gimme/versions` didn't already exist in a place where the new code assumed that it did.

The sort versions are handled: either `sort` supports `--version-sort`, or we fall back to `--general-numeric-sort` after zero-padding any digit sequences after a dot, to make the general-numeric-sort work.

If there's a `sort(1)` which doesn't handle `--general-numeric-sort` then that's an issue to care about.  But while it's a long-opt and not POSIX, the modern BSD reimplementations of `sort` make sure to include that flag for compatibility.  It's one of the GNUisms which is now "standard" on any maintained platform.  For other platforms, we _might_ want to look for a `gsort` command?  But that's a case of "if it's really an issue, fix it".

File: gimme
Comment (philpennock) [reply gimme:107]: (resolved, different issue)

**Review by meatballhat — APPROVED** [review]
:saxophone: :bear:

**Pull request merged**
